LogStreamTest.shouldCloseLogStream unstabled

**Description**

Failed sometimes in the CI.


```
[ERROR] Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.074 s <<< FAILURE! - in io.zeebe.logstreams.log.LogStreamTest

[ERROR] io.zeebe.logstreams.log.LogStreamTest.shouldCloseLogStream  Time elapsed: 0.683 s  <<< FAILURE!

java.lang.AssertionError: 



Expecting code to raise a throwable.

	at io.zeebe.logstreams.log.LogStreamTest.shouldCloseLogStream(LogStreamTest.java:91)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

	at java.base/java.lang.reflect.Method.invoke(Method.java:566)

	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)

	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)

	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)

	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)

	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)

	at org.junit.rules.RunRules.evaluate(RunRules.java:20)

	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)

	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)

	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)

	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

	at org.junit.runners.Suite.runChild(Suite.java:128)

	at org.junit.runners.Suite.runChild(Suite.java:27)

	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)

	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)

	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)

	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)

	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)

	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)
```

Output:

```
12:13:49.119 [] [main] INFO  io.zeebe.test - Test finished: shouldCreateNewLogStreamBatchWriter(io.zeebe.logstreams.log.LogStreamTest)
12:13:49.120 [] [main] INFO  io.zeebe.test - Test started: shouldCloseLogStream(io.zeebe.logstreams.log.LogStreamTest)
12:13:49.313 [io.zeebe.logstreams.impl.LogStreamBuilder$1] [-zb-actors-3] WARN  io.zeebe.logstreams - Unexpected non-empty log failed to read the last block
12:13:49.318 [io.zeebe.logstreams.impl.log.LogStreamImpl] [-zb-actors-3] WARN  io.zeebe.logstreams - Unexpected non-empty log failed to read the last block
12:13:49.533 [io.zeebe.logstreams.impl.log.LogStreamImpl] [-zb-actors-1] DEBUG io.zeebe.logstreams - Configured log appender back pressure at partition 0 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, alphaLimit=0.7, betaLimit=0.95}. Window limiting is disabled
12:13:49.600 [io.zeebe.logstreams.impl.log.LogStreamImpl] [-zb-actors-3] INFO  io.zeebe.logstreams - Close appender for log stream 0
12:13:49.601 [0-write-buffer] [-zb-actors-3] DEBUG io.zeebe.dispatcher - Dispatcher closed
12:13:49.602 [io.zeebe.logstreams.impl.log.LogStreamImpl] [-zb-actors-1] INFO  io.zeebe.logstreams - On closing logstream 0 close 1 readers
12:13:49.603 [io.zeebe.logstreams.impl.log.LogStreamImpl] [-zb-actors-1] INFO  io.zeebe.logstreams - Close log storage with name 0
```

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java
@@ -43,10 +43,12 @@ public class LogStreamImpl extends Actor implements LogStream, AutoCloseable {
   private final List<LogStreamReader> readers;
   private final LogStreamReaderImpl reader;
   private final LogStorage logStorage;
+  private final CompletableActorFuture<Void> closeFuture;
   private ActorFuture<LogStorageAppender> appenderFuture;
   private Dispatcher writeBuffer;
   private LogStorageAppender appender;
   private long commitPosition;
+  private Throwable closeError; // set if any error occurred during closeAsync
 
   public LogStreamImpl(
       final ActorScheduler actorScheduler,
@@ -61,6 +63,7 @@ public class LogStreamImpl extends Actor implements LogStream, AutoCloseable {
     this.partitionId = partitionId;
     this.maxFrameLength = maxFrameLength;
     this.logStorage = logStorage;
+    this.closeFuture = new CompletableActorFuture<>();
 
     try {
       logStorage.open();
@@ -68,10 +71,11 @@ public class LogStreamImpl extends Actor implements LogStream, AutoCloseable {
       throw new UncheckedIOException(e);
     }
 
-    commitPosition = INVALID_ADDRESS;
-    readers = new ArrayList<>();
+    this.commitPosition = INVALID_ADDRESS;
+    this.readers = new ArrayList<>();
     this.reader = new LogStreamReaderImpl(logStorage);
-    readers.add(reader);
+    this.readers.add(reader);
+
     internalSetCommitPosition(reader.seekToEnd());
   }
 
@@ -91,30 +95,36 @@ public class LogStreamImpl extends Actor implements LogStream, AutoCloseable {
   }
 
   @Override
+  protected void onActorClosing() {
+    LOG.info("On closing logstream {} close {} readers", logName, readers.size());
+    readers.forEach(LogStreamReader::close);
+    LOG.info("Close log storage with name {}", logName);
+    logStorage.close();
+  }
+
+  @Override
+  protected void onActorClosed() {
+    if (closeError != null) {
+      closeFuture.completeExceptionally(closeError);
+    } else {
+      closeFuture.complete(null);
+    }
+  }
+
+  @Override
   public ActorFuture<Void> closeAsync() {
     if (actor.isClosed()) {
-      return CompletableActorFuture.completed(null);
+      return closeFuture;
     }
 
-    final var closeFuture = new CompletableActorFuture<Void>();
     actor.call(
-        () -> {
-          closeAppender()
-              .onComplete(
-                  (v, t) -> {
-                    if (t == null) {
-                      LOG.info("On closing logstream {} close {} readers", logName, readers.size());
-                      readers.forEach(LogStreamReader::close);
-
-                      LOG.info("Close log storage with name {}", logName);
-                      logStorage.close();
-                      closeFuture.complete(null);
+        () ->
+            closeAppender()
+                .onComplete(
+                    (nothing, appenderError) -> {
+                      closeError = appenderError;
                       actor.close();
-                    } else {
-                      closeFuture.completeExceptionally(t);
-                    }
-                  });
-        });
+                    }));
     return closeFuture;
   }
 
@@ -125,10 +135,7 @@ public class LogStreamImpl extends Actor implements LogStream, AutoCloseable {
 
   @Override
   public void setCommitPosition(final long commitPosition) {
-    actor.call(
-        () -> {
-          internalSetCommitPosition(commitPosition);
-        });
+    actor.call(() -> internalSetCommitPosition(commitPosition));
   }
 
   private void internalSetCommitPosition(long commitPosition) {

